jni/src/common/Foreign/JNI/Unsafe/Introspection.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+{-# OPTIONS_GHC -fno-warn-name-shadowing #-}
+
+module Foreign.JNI.Unsafe.Introspection
+ ( getClassName
+ , getSignatures
+ , toString
+ , isInstanceOf
+ , classGetNameMethod
+ ) where
+
+import Control.Monad (forM)
+import Data.Foldable (fold)
+import Data.Maybe (catMaybes)
+import Data.ByteString.Builder (toLazyByteString, word16LE)
+import Data.ByteString.Lazy (toStrict)
+import Data.Singletons
+import Data.Text (unpack)
+import Data.Text.Encoding (decodeUtf16LE)
+import Foreign.JNI.Unsafe.Internal
+import Foreign.JNI.Types
+import qualified Foreign.JNI.String as JNI
+import Foreign.Marshal.Array
+import Prelude hiding (String)
+
+-- | The "Class" class.
+kclass :: IO JClass
+kclass = findClass $ referenceTypeName $ sing @('Class "java.lang.Class")
```

Comment:
kclass and kmethod are looking the classes over an over, always yielding the same class.

So far we deal with this in other places with

```Haskell
{#- NOINLINE kclass -#}
kclass :: JClass
kclass = unsafePerformIO $ do
    j <- findClass $ referenceTypeName $ sing @('Class "java.lang.Class")
    newGlobalRef j <* deleteLocalRef j
```

MethodIDs can be defined at the top-level in the same way:
```Haskell
{#- NOINLINE methodGetNameMethod -#}
methodGetNameMethod :: JMethodID
methodGetNameMethod =
    unsafePerformIO $ do
      let sig = methodSignature [] (sing @('Class "java.lang.String"))
      getMethodID kmethod (JNI.fromChars "getName") sig
```

Take a look at http://hackage.haskell.org/package/base-4.14.0.0/docs/System-IO-Unsafe.html

Reply:
Done